### ISSUE TITLE
lxc copy doesn't warn when versions are incompatible.

### ISSUE THREAD

**Issue by abentley** [lxc/lxd#3925]
# Required information

 * Distribution: Ubuntu
 * Distribution version: Xenial
 * The output of "lxc info" or if that fails:
 lxc info
config:
  core.https_address: '[::]:8443'
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses:
  - 192.168.1.13:8443
  - '[fdf0:3b88:9ce6::8d1]:8443'
  - '[2607:f2c0:94f0:3500::8d1]:8443'
  - '[fdf0:3b88:9ce6:0:257a:a72:b4c3:c3b8]:8443'
  - '[fdf0:3b88:9ce6:0:7c35:7426:53d9:89ad]:8443'
  - '[2607:f2c0:94f0:3500:257a:a72:b4c3:c3b8]:8443'
  - '[2607:f2c0:94f0:3500:c07b:fd41:be77:73e7]:8443'
  - 10.0.3.1:8443
  - 10.3.198.1:8443
  - 172.18.0.1:8443
  - 172.17.0.1:8443
  - 192.168.122.1:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIGyjCCBLKgAwIBAgIRALNZdqvV6bhS/M4lZRcmpiMwDQYJKoZIhvcNAQELBQAw
    NDEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEUMBIGA1UEAwwLcm9vdEBz
    a2lubnkwHhcNMTYwMjExMjIwOTU4WhcNMjYwMjA4MjIwOTU4WjA0MRwwGgYDVQQK
    ExNsaW51eGNvbnRhaW5lcnMub3JnMRQwEgYDVQQDDAtyb290QHNraW5ueTCCAiIw
    DQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBALZgID0ev+ifq5r7Jj1mvjDMAY17
    JfuWP0eX5ECnclmYeO5O/m9J5+N97R1q3BqGh7puCqeZsuSPdlrkrioGNhtKJKGY
    YDP4MMArOi5cZ0GBUiuGwQJM22AhI3QjgA/BjSus7Ppqud3nhfn830OGfRNnXPm+
    meS9TMO5cV80ZDhGelqcws9vHhCx+RYs87oUjPQqAXIilxS85mgaXR6w5+JzUGKl
    NIP6u+Yn6+4FIh6ueqh2xeFsgGX79HkWB1smEFfq4+p2F443bcPGpTVxbEDcfAQk
    FQbMap4PXn4AR9RFOl5f2b/Q21SmOf5WOr+X29dRq6WKa5HKBggjwzGSe+IU00/D
    XKguRdYxrrf5TVANGdCiZb7nneySlNAtwznSlVzM6jUB6ptB2fJ5rev03+32s6nh
    DfODiWA5xFuqrNKMtNZsSqovbBRk6Bg3Uu9JBKIwD/FoiemgtS+D4tjqNaiMqXx8
    8fdpgNyl4ZLLKbSytHd8dREHHIllZU0L4hjMOASqLdjPC51IusszHSZu4HDbLRcS
    jBHypaAjg2R9UKcFWIZorRrAov239iCwhzMsT4LJjZXRSLZ+lyIiCePzvQLj+WW/
    u7aK/SqNKXNFaDDHannm0Qo/9ji07KJVMgHyHwoimOzcw4mPWJGzA+1KxtxUpxs2
    9bQaCP0BkIdNUY4ZAgMBAAGjggHVMIIB0TAOBgNVHQ8BAf8EBAMCBaAwEwYDVR0l
    BAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADCCAZoGA1UdEQSCAZEwggGNggZz
    a2lubnmCDzE5Mi4xNjguMS4xMy8yNIInZmRmMDozYjg4OjljZTY6MDpmOWU5OmI2
    ODY6N2RlOTpmYzRhLzY0gidmZGYwOjNiODg6OWNlNjowOjVlNTE6NGZmZjpmZTc1
    OmQ4NTIvNjSCKjI2MDc6ZjJjMDo4MDA4OmRkOWY6ZjllOTpiNjg2OjdkZTk6ZmM0
    YS82NIIqMjYwNzpmMmMwOjgwMDg6ZGQ5Zjo1ZTUxOjRmZmY6ZmU3NTpkODUyLzY0
    ghxmZTgwOjo1ZTUxOjRmZmY6ZmU3NTpkODUyLzY0ghAxOTIuMTY4LjEyMi4xLzI0
    ggsxMC4wLjMuMS8yNIIcZmU4MDo6MWM3MDphZGZmOmZlYmU6ZDhjMi82NIIcZmU4
    MDo6ZmNmMTo2ZWZmOmZlNTU6MmExMS82NIIcZmU4MDo6ZmNlZjphOGZmOmZlOGM6
    Yjk5OS82NIIQMTAuMTcyLjY1LjIzNi8xOIIfMjAwMTo2N2M6MTU2Mjo4MDA3Ojph
    YWM6NDFlYy82NDANBgkqhkiG9w0BAQsFAAOCAgEAL2LyZhwZiC5sAsrUTTbzbFuF
    mIhnJ1D6NJxX9nuamzkFXaMo119bHVfwiz6k72lK3zhGBSwgCmwTjuXPSsItTs+Y
    d8myU7seYzPwF5dvuvra5LNvgJkqfUHcjWYvHD55Gavvh5yvP0HTcS4quUNCzgBS
    2AsOHlBYmXdmAvFn8Y09CC0cXLEyJJ9DWEliSIdJIE0CB7Wdfs/bU0ynTVGo8PVr
    eNrR2ldKp8oVz7zHn0Wpvees4kESNeuq5EzyrCwLUsdq3W8IsZE351qVij5hzjhO
    8QWh2KqpEBHO/g0LIj6w5ZwfsRjRkz/WwyN/Jx0h/LEiZscqqERw7+wH8JCSrQwj
    xsZc26NwGuiUta8NMqZBIVczW7vf/533O+5T6rcqkcxYR9lzF8oqsNK+tDKAB99d
    xIvQzd57WeC/brjG+t7RKR4LgMNSoyMWmG5qqHve9l/KXB2ay93JuFdFQXXc6hAc
    VCZv0YaKGfo0jEmGlYfIy9mhmKeeEeH+o0Y9kdW2ESb4SxvCoE3xmhrl1LTtMkuC
    wwKiiLUpYVK4rhElqOC5/osky/77VC6iVTgJ+LXrEhNcP2lMWi3Km84JvUxTYxf6
    aCSYb+5mslNU3UIYPsnDDWcyx3tjJzEYmQjdhHi8u6nZPyqlKE/tqLuKzRADpuG4
    lt00J0Sq4e2XkK2pCAs=
    -----END CERTIFICATE-----
  certificate_fingerprint: 9cd55966bc788d26dee12d36835024f71e372515c2b121ed9b12b58b8a64e072
  driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-96-generic
  server: lxd
  server_pid: 1491
  server_version: "2.18"
  storage: dir
  storage_version: "1"

# Issue description

When two versions are incompatible, and "lxc copy" is issued, a series of "Bad key" messages are given, instead of explaining that the versions are incompatible.  When the incompatible versions are 2.18 and 2.10, the output looks like:
 - https://192.168.1.13:8443: Bad key: image.serial
 - https://[fdf0:3b88:9ce6::8d1]:8443: Bad key: image.description
 - https://[2607:f2c0:94f0:3500::8d1]:8443: Bad key: image.description
 - https://[fdf0:3b88:9ce6:0:257a:a72:b4c3:c3b8]:8443: Bad key: image.description
 - https://[fdf0:3b88:9ce6:0:7c35:7426:53d9:89ad]:8443: Bad key: image.serial
 - https://[2607:f2c0:94f0:3500:257a:a72:b4c3:c3b8]:8443: Bad key: image.release
 - https://[2607:f2c0:94f0:3500:c07b:fd41:be77:73e7]:8443: Bad key: image.serial
 - https://10.0.3.1:8443: Bad key: image.label
 - https://10.3.198.1:8443: Bad key: image.version
 - https://172.18.0.1:8443: Bad key: image.version
 - https://172.17.0.1:8443: Bad key: image.architecture
 - https://192.168.122.1:8443: Bad key: image.release

# Steps to reproduce

 1. Get an Ubuntu Xenial machine.  Install 2.1.8 from backports.
 2. Get a second Ubuntu Xenial machine.  Install 2.0.10, the default.
 3. Create a container on the first machine
 4. On the first machine, add the second machine as a remote.
 5. On the first machine, use "lxc copy" to copy the local container to the remote machine (e.g. lxc copy xavier machinetwo:xavier)

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by stgraber** [comment]
Well, the target LXD does tell you what config key it doesn't like. Clearing those from your source container's config with "lxc config edit" should then make the copy succeed.

There aren't fundamental incompatibilities between LXD releases as we don't break our API, but config keys for new features will be rejected by older versions. It's then up to you whether you're willing to "downgrade" your container by unsetting the keys in question.

**Comment by stgraber** [comment]
The crazy long output is a bit unfortunate, I'll have to take a look to see if the initial config validation can be done before we hit the remote handling code which causes the query to be attempted through all known network paths.

**Comment by abentley** [comment]
I understand what you're saying about the API, but I don't think it's completely true.  You have a schema for config keys, and that is an implicit piece of the API.  By changing the schema in 2.1.18, you're producing API calls that are incompatible with the API as implemented in 2.0.10.

An error about a version mismatch would have saved me a lot of time.  I didn't know about the version mismatch initially.  I assumed both were running the same version, since both were xenial machines and had lxd installed from the ubuntu archive and not a ppa.

I don't know anything about config keys.  I did not create any of them, and I don't know which ones are mandatory or optional in various lxd versions.  It's not user-friendly to ask me to decide whether I'm willing to downgrade my container by removing keys, because I never chose to set them and I don't know what they mean.

Since it looked like a version mismatch to me, I just installed 2.1.18 on the second machine.

Also, "bad key" may be too strong.  That suggests that there's something _wrong_ with the container, not that it's unsupported.  May I suggest "unknown key" instead?
